### DISCUSSION TITLE
tools/webplotdigitizer/

### DISCUSSION THREAD

**Discussion by giscus** [kausalflow/tools#115] [Comments]
# tools/webplotdigitizer/

Web based tool to extract data from plots, images, and maps

https://tools.kausalflow.com/tools/webplotdigitizer/

**Comment by cerberuszzc** [comment]
import json
from datetime import datetime, timedelta
import webplotting as wp
import pandas as pd
from statsmodels.tsa.arima.model import ARIMA

def parse_date(date_str):
    try:
        return datetime.strptime(date_str, "%Y-%j").date()
    except ValueError:
        return date_str

def load_historical_data():
    """加载最新的历史数据"""
    url = 'https://api.thingspeak.com/v2/price chart/?taid=150176&interval=day'
    try:
        data = requests.get(url).json()
        time_series = []
        dates = []
        for record in data:
            date_str = record['time']
            try:
                date = parse_date(date_str)
                exchange_rate = float(record['value'])
                time_series.append(exchange_rate)
                dates.append(date)
            except ValueError as e:
                print(f"记录{record}：错误日期格式。")
        return pd.DataFrame(time_series), dates
    except Exception as e:
        print(f"获取最新数据失败：{e}")
        return None, None

def load_daily_data():
    """加载日元对澳元的最新数据"""
    url = 'https://api.thingspeak.com/v2/price chart/?taid=150178&interval=day'
    try:
        data = requests.get(url).json()
        time_series_dz = []
        dates_dz = []
        for record in data:
            date_str = record['time']
            try:
                date = parse_date(date_str)
                exchange_rate = float(record['value'])
                time_series_dz.append(exchange_rate)
                dates_dz.append(date)
            except ValueError as e:
                print(f"记录{record}：错误日期格式。")
        return pd.DataFrame({'Date': dates_dz}, index=dates_dz), dates_dz
    except Exception as e:
        print(f"获取最新日元对澳元数据失败：{e}")
        return None, None

def get_realtime预测未来澳元对日元汇率：
    """
    实现实时预测澳元对日元的汇率并生成趋势图。
    
    提供未来3天的预测结果，并与实际历史数据进行对比分析。
    使用 webPlotting 以获取动态图表支持。
    """
    def load_historical_data():
        url = 'https://api.thingspeak.com/v2/price chart/?taid=150176&interval=day'
        try:
            data = requests.get(url).json()
            time_series_au, dates = data['ExchangeRate']
            time_series_au = [float(rate) for rate in time_series_au]
            return pd.DataFrame({'Date': dates}, index=dates), dates
        except Exception as e:
            print(f"获取最新澳元对日元历史数据失败：{e}")
            return None, None

    def load_daily_data():
        url = 'https://api.thingspeak.com/v2/price chart/?taid=150178&interval=day'
        try:
            data = requests.get(url).json()
            time_series_dz, dates_dz = data['ExchangeRate']
            time_series_dz = [float(rate) for rate in time_series_dz]
            return pd.DataFrame({'Date': dates_dz}, index=dates_dz), dates_dz
        except Exception as e:
            print(f"获取最新日元对澳元数据失败：{e}")
            return None, None

    try:
        # 提取最新的历史数据和日元对澳元的历史数据
        au_data, _ = load_historical_data()
        dz_data, date_list_dz = load_daily_data()

        if len(au_data) >= 4 and len(dz_data) >= 4:
            # 提取时间序列并排序
            merged_au = []
            for date, rate in zip(date_list_dz, au_data):
                merged_au.append((date, rate))
            merged_au_df = pd.DataFrame(merged_au, columns=['Date', 'ExchangeRate'])
            merged_au_df.sort_values(by='Date', inplace=True)

            # 计算平均汇率和标准差
            mean_au = merged_au_df['ExchangeRate'].mean()
            std_au = merged_au_df['ExchangeRate'].std()

            # 提取时间序列并转换为Series
            exchange_rate_series_au = S(merged_au_df['ExchangeRate'])

            # 选择合适的ARIMA模型参数（优化阶数）
            try:
                order = (1, 0, 2)  # 预设的最小阶数组合
                model = ARIMA(exchange_rate_series_au, order=order)
                model_fit = model.fit()

                # 预测未来3天的汇率
                forecast_steps = 3
                try:
                    forecast = model_fit.forecast(steps=forecast_steps)
                    predictedexchangeRates = [rate for rate in forecast]
                except Exception as e:
                    print(f"模型拟合失败：{e}")
                    return None

                # 绘制实际数据和预测结果的汇率趋势图
                actualExchangeRate = S(au_data['ExchangeRate'])
                actualExchangeRate.plot()

                forecastExchangeRate = pd.Series(predictedexchangeRates)
                forecastExchangeRate.plot(alpha=0.5, color='g')

                plt.xlabel('时间')
                plt.ylabel('汇率（人民币）')
                plt.title('澳元对日元的历史和未来3天的汇率预测')
                plt.legend()
                plt.grid(True)
                plt.tight_layout()

                plt.show()
            except Exception as e:
                print(f"模型预测失败：{e}")
                return None
        else:
            print("需要至少4个最新的历史数据（澳元对日元）才能进行预测。")
            return None

    # 自愈机制处理异常情况
    except Exception as e:
        print(f"代码在运行时遇到异常：{str(e)}\n自愈机制启用，继续执行...")
        return wp.new_graph()

    finally:
        pass

if __name__ == "__main__":
    main()

**Reply by cerberuszzc** [reply]
2025.2.6
